File: packages/agent-workspace/src/workflow/builtin.ts
import type { WorkflowDef } from "./types";

/**
 * The builtin fallback workflow — the intake router's DEFAULT ROUTE, not a
 * dumping ground. Its one job: do the work and, when finished, signal done
 * (which admits the terminal stage). Specific workflows are registered on top;
 * anything that doesn't match one flows through here.
 */
export const GENERAL_WORKFLOW: WorkflowDef = {
  id: "general",
  version: "1",
  name: "General",
  description: "Fallback workflow for any task without a specific one.",
  fields: {
    request: { type: "string", description: "The original requirement / what was asked." },
    summary: { type: "string", description: "One-line outcome, written when finishing." },
  },
  stages: [
    {
      id: "open",
      category: "in_progress",
      entry: { op: "always" },
      instructions:
        "Do whatever the task needs. Record a one-line `summary` of the outcome, then request a transition to close it.",
    },
    {
      id: "done",
      category: "done",
      // Admitted once the agent explicitly signals it is finished.
      entry: { op: "hasEvent", eventType: "transition.requested" },
    },
  ],
};

File: internal/workflow/builtin.go
package workflow

// GeneralWorkflow returns the built-in fallback workflow.
func GeneralWorkflow() WorkflowDef {
	return WorkflowDef{
		ID:          "general",
		Version:     "1",
		Name:        "General",
		Description: "Fallback workflow for any task without a specific one.",
		Fields: map[string]FieldDef{
			"request": {Type: "string", Description: "The original requirement / what was asked."},
			"summary": {Type: "string", Description: "One-line outcome, written when finishing."},
		},
		Stages: []StageDef{
			{
				ID:           "open",
				Category:     StageCategoryInProgress,
				Entry:        Guard{Op: "always"},
				Effort:       "medium",
				OutputFields: []string{"summary"},
				Instructions: "Do whatever the task needs. Record a one-line summary of the outcome, then request a transition to close it.",
			},
			{
				ID:       "done",
				Category: StageCategoryDone,
				Entry:    Guard{Op: "hasEvent", EventType: TaskEventTransitionRequested},
			},
		},
	}
}
